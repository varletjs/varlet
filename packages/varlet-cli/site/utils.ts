import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
