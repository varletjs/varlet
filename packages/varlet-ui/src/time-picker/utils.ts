import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
