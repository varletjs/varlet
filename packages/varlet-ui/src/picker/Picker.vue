<template>
  <component
    :is="dynamic ? n('$-popup') : Transition"
    v-bind="
      dynamic
        ? {
            onOpen,
            onOpened,
            onClose,
            onClosed,
            onClickOverlay,
            onRouteChange,
            closeOnClickOverlay,
            teleport,
            show,
            'onUpdate:show': handlePopupUpdateShow,
            position: 'bottom',
            class: n('popup'),
          }
        : null
    "
    var-picker-cover
  >
    <div :class="n()" v-bind="$attrs">
      <div :class="n('toolbar')" v-if="toolbar">
        <slot name="cancel">
          <var-button
            :class="n('cancel-button')"
            var-picker-cover
            text
            :text-color="cancelButtonTextColor"
            @click="cancel"
          >
            {{ dt(cancelButtonText, pack.pickerCancelButtonText) }}
          </var-button>
        </slot>
        <slot name="title">
          <div :class="n('title')">{{ dt(title, pack.pickerTitle) }}</div>
        </slot>
        <slot name="confirm">
          <var-button
            :class="n('confirm-button')"
            text
            var-picker-cover
            :text-color="confirmButtonTextColor"
            @click="confirm"
          >
            {{ dt(confirmButtonText, pack.pickerConfirmButtonText) }}
          </var-button>
        </slot>
      </div>
      <div :class="n('columns')" :style="{ height: `${columnHeight}px` }">
        <div
          :class="n('column')"
          v-for="c in scrollColumns"
          :key="c.id"
          @touchstart="handleTouchstart($event, c)"
          @touchmove.prevent="handleTouchmove($event, c)"
          @touchend="handleTouchend($event, c)"
        >
          <div
            :class="n('scroller')"
            :ref="(el) => getScrollEl(el, c)"
            :style="{
              transform: `translateY(${c.translate}px)`,
              transitionDuration: `${c.duration}ms`,
              transitionProperty: c.duration ? 'transform' : 'none',
            }"
            @transitionend="handleTransitionend(c)"
          >
            <div :class="n('option')" :style="{ height: `${optionHeight}px` }" v-for="t in c.column.texts" :key="t">
              <div :class="n('text')">{{ textFormatter(t, c.columnIndex) }}</div>
            </div>
          </div>
        </div>
        <div
          :class="n('picked')"
          :style="{
            top: `${center}px`,
            height: `${optionHeight}px`,
          }"
        ></div>
        <div :class="n('mask')" :style="{ backgroundSize: `100% ${(columnHeight - optionHeight) / 2}px` }"></div>
      </div>
    </div>
  </component>
</template>

<script lang="ts">
import VarButton from '../button'
import VarPopup from '../popup'
import { defineComponent, watch, ref, computed, Transition, toRaw } from 'vue'
import { props } from './props'
import { isArray } from '@varlet/shared'
import { dt } from '../utils/shared'
import { toPxNum, getTranslate } from '../utils/elements'
import { pack } from '../locale'
import type { Ref, ComputedRef, ComponentPublicInstance } from 'vue'
import type { CascadeColumn, NormalColumn } from './props'
import type { Texts } from './index'
import { createNamespace, call } from '../utils/components'

export interface ScrollColumn {
  id: number
  touching: boolean
  index: number
  columnIndex: number
  prevY: number | undefined
  momentumPrevY: number | undefined
  momentumTime: number
  translate: number
  duration: number
  scrolling: boolean
  column: NormalColumn
  columns?: CascadeColumn[]
  scrollEl: HTMLElement | null
}

const { n, classes } = createNamespace('picker')

const MOMENTUM_RECORD_TIME = 300
const MOMENTUM_ALLOW_DISTANCE = 15

let sid = 0

export default defineComponent({
  name: 'VarPicker',
  components: {
    VarButton,
    VarPopup,
  },
  inheritAttrs: false,
  props,
  setup(props) {
    const scrollColumns: Ref<ScrollColumn[]> = ref([])
    const optionHeight: ComputedRef<number> = computed(() => toPxNum(props.optionHeight))
    const optionCount: ComputedRef<number> = computed(() => toPxNum(props.optionCount))
    const center: ComputedRef<number> = computed(
      () => (optionCount.value * optionHeight.value) / 2 - optionHeight.value / 2
    )
    const columnHeight: ComputedRef<number> = computed(() => optionCount.value * optionHeight.value)
    let prevIndexes: number[] = []

    const getScrollEl = (el: Element | ComponentPublicInstance | null, scrollColumn: ScrollColumn) => {
      scrollColumn.scrollEl = el as HTMLElement
    }

    const handlePopupUpdateShow = (value: boolean) => {
      call(props['onUpdate:show'], value)
    }

    const limitTranslate = (scrollColumn: ScrollColumn) => {
      const START_LIMIT = optionHeight.value + center.value
      const END_LIMIT = center.value - scrollColumn.column.texts.length * optionHeight.value

      if (scrollColumn.translate >= START_LIMIT) {
        scrollColumn.translate = START_LIMIT
      }
      if (scrollColumn.translate <= END_LIMIT) {
        scrollColumn.translate = END_LIMIT
      }
    }

    const boundaryIndex = (scrollColumn: ScrollColumn, index: number) => {
      const { length } = scrollColumn.column.texts

      index = index >= length ? length - 1 : index
      index = index <= 0 ? 0 : index

      return index
    }

    const getIndex = (scrollColumn: ScrollColumn) => {
      const index = Math.round((center.value - scrollColumn.translate) / optionHeight.value)

      return boundaryIndex(scrollColumn, index)
    }

    const getPicked = () => {
      const texts = scrollColumns.value.map((scrollColumn) => scrollColumn.column.texts[scrollColumn.index])
      const indexes = scrollColumns.value.map((scrollColumn) => scrollColumn.index)

      return {
        texts,
        indexes,
      }
    }

    const scrollTo = (scrollColumn: ScrollColumn, index: number, duration: number, noEmit = false) => {
      const translate = center.value - boundaryIndex(scrollColumn, index) * optionHeight.value

      if (translate === scrollColumn.translate) {
        scrollColumn.scrolling = false
        !noEmit && change(scrollColumn)
      }

      scrollColumn.translate = translate
      scrollColumn.duration = duration
    }

    const momentum = (scrollColumn: ScrollColumn, distance: number, duration: number) => {
      scrollColumn.translate += (Math.abs(distance / duration) / 0.003) * (distance < 0 ? -1 : 1)
    }

    const handleTouchstart = (event: TouchEvent, scrollColumn: ScrollColumn) => {
      scrollColumn.touching = true
      scrollColumn.scrolling = false
      scrollColumn.duration = 0
      scrollColumn.translate = getTranslate(scrollColumn.scrollEl as HTMLElement)
    }

    const handleTouchmove = (event: TouchEvent, scrollColumn: ScrollColumn) => {
      if (!scrollColumn.touching) {
        return
      }

      const { clientY } = event.touches[0]
      const moveY = scrollColumn.prevY !== undefined ? clientY - scrollColumn.prevY : 0
      scrollColumn.prevY = clientY
      scrollColumn.translate += moveY

      limitTranslate(scrollColumn)

      const now = performance.now()
      if (now - scrollColumn.momentumTime > MOMENTUM_RECORD_TIME) {
        scrollColumn.momentumTime = now
        scrollColumn.momentumPrevY = scrollColumn.translate
      }
    }

    const handleTouchend = (event: TouchEvent, scrollColumn: ScrollColumn) => {
      scrollColumn.touching = false
      scrollColumn.scrolling = true
      scrollColumn.prevY = undefined
      const distance = scrollColumn.translate - (scrollColumn.momentumPrevY as number)
      const duration = performance.now() - scrollColumn.momentumTime
      const shouldMomentum = Math.abs(distance) >= MOMENTUM_ALLOW_DISTANCE && duration <= MOMENTUM_RECORD_TIME

      shouldMomentum && momentum(scrollColumn, distance, duration)

      scrollColumn.index = getIndex(scrollColumn)
      scrollTo(scrollColumn, scrollColumn.index, shouldMomentum ? 1000 : 200)
    }

    const handleTransitionend = (scrollColumn: ScrollColumn) => {
      scrollColumn.scrolling = false
      change(scrollColumn)
    }

    const normalizeNormalColumns = (normalColumns: NormalColumn[]) => {
      return normalColumns.map((column: NormalColumn | any[], columnIndex: number) => {
        const normalColumn = (isArray(column) ? { texts: column } : column) as NormalColumn
        const scrollColumn: ScrollColumn = {
          id: sid++,
          prevY: undefined,
          momentumPrevY: undefined,
          touching: false,
          translate: center.value,
          index: normalColumn.initialIndex ?? 0,
          columnIndex,
          duration: 0,
          momentumTime: 0,
          column: normalColumn,
          scrollEl: null,
          scrolling: false,
        }
        scrollTo(scrollColumn, scrollColumn.index, 0, true)
        return scrollColumn
      })
    }

    const normalizeCascadeColumns = (cascadeColumns: CascadeColumn[]) => {
      const scrollColumns: ScrollColumn[] = []

      createChildren(scrollColumns, cascadeColumns, 0, true)

      return scrollColumns
    }

    const createChildren = (
      scrollColumns: ScrollColumn[],
      children: CascadeColumn[],
      columnIndex: number,
      initial = false
    ) => {
      if (isArray(children) && children.length) {
        const index = initial ? props.cascadeInitialIndexes[scrollColumns.length] ?? 0 : 0

        const scrollColumn: ScrollColumn = {
          id: sid++,
          prevY: undefined,
          momentumPrevY: undefined,
          touching: false,
          translate: center.value,
          index,
          columnIndex,
          duration: 0,
          momentumTime: 0,
          column: {
            texts: children.map((cascadeColumn) => cascadeColumn[props.textKey]),
          },
          columns: children,
          scrollEl: null,
          scrolling: false,
        }

        scrollColumns.push(scrollColumn)
        scrollTo(scrollColumn, scrollColumn.index, 0, true)
        createChildren(
          scrollColumns,
          (scrollColumn.columns as CascadeColumn[])[scrollColumn.index].children,
          columnIndex + 1,
          initial
        )
      }
    }

    const rebuildChildren = (scrollColumn: ScrollColumn) => {
      scrollColumns.value.splice(scrollColumns.value.indexOf(scrollColumn) + 1)
      createChildren(
        scrollColumns.value,
        (scrollColumn.columns as CascadeColumn[])[scrollColumn.index].children,
        scrollColumn.columnIndex + 1
      )
    }

    const change = (scrollColumn: ScrollColumn) => {
      const { cascade, onChange } = props
      cascade && rebuildChildren(scrollColumn)

      const hasScrolling = scrollColumns.value.some((scrollColumn) => scrollColumn.scrolling)
      if (hasScrolling) {
        return
      }

      const { texts, indexes } = getPicked()

      const samePicked = indexes.every((index, idx) => index === prevIndexes[idx])
      if (samePicked) {
        return
      }

      prevIndexes = [...indexes]
      call(onChange, texts, indexes)
    }

    const stopScroll = () => {
      if (props.cascade) {
        const currentScrollColumn = scrollColumns.value.find((scrollColumn) => scrollColumn.scrolling)
        if (currentScrollColumn) {
          currentScrollColumn.translate = getTranslate(currentScrollColumn.scrollEl as HTMLElement)
          currentScrollColumn.index = getIndex(currentScrollColumn)
          scrollTo(currentScrollColumn, currentScrollColumn.index, 0, true)
          currentScrollColumn.scrolling = false
          rebuildChildren(currentScrollColumn)
        }
      } else {
        scrollColumns.value.forEach((scrollColumn) => {
          scrollColumn.translate = getTranslate(scrollColumn.scrollEl as HTMLElement)
          scrollColumn.index = getIndex(scrollColumn)
          scrollTo(scrollColumn, scrollColumn.index, 0)
        })
      }
    }

    // expose
    const confirm = () => {
      stopScroll()

      const { texts, indexes } = getPicked()
      prevIndexes = [...indexes]
      call(props.onConfirm, texts, indexes)
    }

    // expose
    const cancel = () => {
      stopScroll()

      const { texts, indexes } = getPicked()
      prevIndexes = [...indexes]
      call(props.onCancel, texts, indexes)
    }

    watch(
      () => props.columns,
      (newValue: any) => {
        scrollColumns.value = props.cascade
          ? normalizeCascadeColumns(toRaw(newValue) as CascadeColumn[])
          : normalizeNormalColumns(toRaw(newValue) as NormalColumn[] | Texts)

        const { indexes } = getPicked()
        prevIndexes = [...indexes]
      },
      {
        immediate: true,
        deep: true,
      }
    )

    return {
      n,
      classes,
      pack,
      optionHeight,
      optionCount,
      scrollColumns,
      columnHeight,
      center,
      Transition,
      getScrollEl,
      handlePopupUpdateShow,
      handleTouchstart,
      handleTouchmove,
      handleTouchend,
      handleTransitionend,
      confirm,
      cancel,
      dt,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../button/button';
@import '../popup/popup';
@import './picker';
</style>
