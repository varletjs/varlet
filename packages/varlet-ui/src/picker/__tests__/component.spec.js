import Picker from '..'
import VarPicker from '../Picker'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'
import { delay, mockTranslate, trigger } from '../../utils/jest'

test('test picker component plugin', () => {
  const app = createApp({}).use(Picker.Component)
  expect(app.component(Picker.Component.name)).toBeTruthy()
})

const columns = [['A', 'B', 'C']]

const triggerDrag = async (element, x, y) => {
  await trigger(element, 'touchstart', 0, 0)
  await trigger(element, 'touchmove', x / 4, y / 4)
  await trigger(element, 'touchmove', x / 3, y / 3)
  await trigger(element, 'touchmove', x / 2, y / 2)
  await trigger(element, 'touchmove', x, y)
  await delay(300)
  await trigger(element, 'touchend', x, y)
}

test('test scroll up & onConfirm', async () => {
  const { mockRestore } = mockTranslate()
  const onConfirm = jest.fn()

  const wrapper = mount(VarPicker, {
    props: {
      columns,
      onConfirm,
    },
  })
  expect(wrapper.html()).toMatchSnapshot()
  const { element } = wrapper.find('.var-picker__column')

  await triggerDrag(element, 0, -600)
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.var-picker__confirm-button').trigger('click')
  expect(onConfirm).lastCalledWith(['C'], [2])

  mockRestore()
  wrapper.unmount()
})

test('test scroll down & onCancel', async () => {
  const { mockRestore } = mockTranslate()
  const onCancel = jest.fn()

  const wrapper = mount(VarPicker, {
    props: {
      columns,
      onCancel,
    },
  })
  expect(wrapper.html()).toMatchSnapshot()
  const { element } = wrapper.find('.var-picker__column')

  await triggerDrag(element, 0, 600)
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.var-picker__cancel-button').trigger('click')
  expect(onCancel).lastCalledWith(['A'], [0])

  mockRestore()
  wrapper.unmount()
})

test('test confirm & cancel method', async () => {
  const { mockRestore } = mockTranslate()
  const onCancel = jest.fn()
  const onConfirm = jest.fn()

  const wrapper = mount(VarPicker, {
    props: {
      columns,
      onConfirm,
      onCancel,
    },
  })

  wrapper.vm.confirm()
  wrapper.vm.cancel()

  expect(onCancel).lastCalledWith(['A'], [0])
  expect(onConfirm).lastCalledWith(['A'], [0])

  mockRestore()
  wrapper.unmount()
})

test('test cascade mode', async () => {
  const { mockRestore } = mockTranslate()
  const onConfirm = jest.fn()

  const wrapper = mount(VarPicker, {
    props: {
      cascade: true,
      columns: [
        {
          text: '成都市',
          children: [
            {
              text: '温江区',
            },
            {
              text: '金牛区',
            },
          ],
        },
        {
          text: '无锡市',
          children: [
            {
              text: '新吴区',
            },
            {
              text: '惠山区',
            },
          ],
        },
      ],
      onConfirm,
    },
  })

  let columns = wrapper.findAll('.var-picker__column')
  expect(wrapper.html()).toMatchSnapshot()
  expect(columns.length).toBe(2)

  const confirmButton = wrapper.find('.var-picker__confirm-button')

  await triggerDrag(columns[0].element, 0, -44)
  await confirmButton.trigger('click')
  expect(onConfirm).lastCalledWith(['无锡市', '新吴区'], [1, 0])

  // find elements again for rebuild children
  columns = wrapper.findAll('.var-picker__column')
  await triggerDrag(columns[1].element, 0, -44)
  await confirmButton.trigger('click')
  expect(onConfirm).lastCalledWith(['无锡市', '惠山区'], [1, 1])

  mockRestore()
  wrapper.unmount()
})

test('test cascade initial indexes', async () => {
  const onConfirm = jest.fn()
  const onCancel = jest.fn()

  const wrapper = mount(VarPicker, {
    props: {
      cascade: true,
      cascadeInitialIndexes: [1, 0],
      columns: [
        {
          text: '成都市',
          children: [
            {
              text: '温江区',
            },
            {
              text: '金牛区',
            },
          ],
        },
        {
          text: '无锡市',
          children: [
            {
              text: '新吴区',
            },
            {
              text: '惠山区',
            },
          ],
        },
      ],
      onConfirm,
      onCancel,
    },
  })

  wrapper.vm.confirm()
  wrapper.vm.cancel()

  expect(onCancel).lastCalledWith(['无锡市', '新吴区'], [1, 0])
  expect(onConfirm).lastCalledWith(['无锡市', '新吴区'], [1, 0])

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test picker component textFormatter', async () => {
  const textFormatter = jest.fn().mockReturnValue('text')

  const wrapper = mount(VarPicker, {
    props: {
      columns,
      textFormatter,
    },
  })

  const pickerText = wrapper.find('.var-picker__text')

  expect(pickerText.text()).toBe('text')

  wrapper.unmount()
})
