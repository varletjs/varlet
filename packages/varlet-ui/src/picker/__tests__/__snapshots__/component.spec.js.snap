// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test cascade initial indexes 1`] = `
"<transition-stub>
  <div class=\\"var-picker\\">
    <div class=\\"var-picker__toolbar\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__cancel-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">取消</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
      <div class=\\"var-picker__title\\">请选择</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__confirm-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">确认</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
    </div>
    <div class=\\"var-picker__columns\\" style=\\"height: 264px;\\">
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(66px); transition-duration: 0ms; transition-property: none;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">成都市</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">无锡市</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(110px); transition-duration: 0ms; transition-property: none;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">新吴区</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">惠山区</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
      <div class=\\"var-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
    </div>
  </div>
</transition-stub>"
`;

exports[`test cascade mode 1`] = `
"<transition-stub>
  <div class=\\"var-picker\\">
    <div class=\\"var-picker__toolbar\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__cancel-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">取消</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
      <div class=\\"var-picker__title\\">请选择</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__confirm-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">确认</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
    </div>
    <div class=\\"var-picker__columns\\" style=\\"height: 264px;\\">
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(110px); transition-duration: 0ms; transition-property: none;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">成都市</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">无锡市</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(110px); transition-duration: 0ms; transition-property: none;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">温江区</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">金牛区</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
      <div class=\\"var-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
    </div>
  </div>
</transition-stub>"
`;

exports[`test scroll down & onCancel 1`] = `
"<transition-stub>
  <div class=\\"var-picker\\">
    <div class=\\"var-picker__toolbar\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__cancel-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">取消</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
      <div class=\\"var-picker__title\\">请选择</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__confirm-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">确认</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
    </div>
    <div class=\\"var-picker__columns\\" style=\\"height: 264px;\\">
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(110px); transition-duration: 0ms; transition-property: none;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">A</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">B</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">C</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
      <div class=\\"var-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
    </div>
  </div>
</transition-stub>"
`;

exports[`test scroll down & onCancel 2`] = `
"<transition-stub>
  <div class=\\"var-picker\\">
    <div class=\\"var-picker__toolbar\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__cancel-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">取消</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
      <div class=\\"var-picker__title\\">请选择</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__confirm-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">确认</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
    </div>
    <div class=\\"var-picker__columns\\" style=\\"height: 264px;\\">
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(110px); transition-duration: 200ms; transition-property: transform;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">A</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">B</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">C</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
      <div class=\\"var-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
    </div>
  </div>
</transition-stub>"
`;

exports[`test scroll up & onConfirm 1`] = `
"<transition-stub>
  <div class=\\"var-picker\\">
    <div class=\\"var-picker__toolbar\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__cancel-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">取消</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
      <div class=\\"var-picker__title\\">请选择</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__confirm-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">确认</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
    </div>
    <div class=\\"var-picker__columns\\" style=\\"height: 264px;\\">
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(110px); transition-duration: 0ms; transition-property: none;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">A</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">B</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">C</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
      <div class=\\"var-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
    </div>
  </div>
</transition-stub>"
`;

exports[`test scroll up & onConfirm 2`] = `
"<transition-stub>
  <div class=\\"var-picker\\">
    <div class=\\"var-picker__toolbar\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__cancel-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">取消</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
      <div class=\\"var-picker__title\\">请选择</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-picker__confirm-button\\" type=\\"button\\" var-picker-cover=\\"\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">确认</div>
        <div class=\\"var-button__overlay\\"></div>
      </button>
    </div>
    <div class=\\"var-picker__columns\\" style=\\"height: 264px;\\">
      <div class=\\"var-picker__column\\">
        <div class=\\"var-picker__scroller\\" style=\\"transform: translateY(22px); transition-duration: 200ms; transition-property: transform;\\">
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">A</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">B</div>
          </div>
          <div class=\\"var-picker__option\\" style=\\"height: 44px;\\">
            <div class=\\"var-picker__text\\">C</div>
          </div>
        </div>
      </div>
      <div class=\\"var-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
      <div class=\\"var-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
    </div>
  </div>
</transition-stub>"
`;
